### ISSUE TITLE
Lost gradient after clipping in focal loss?

### ISSUE THREAD

**Issue by danijar** [tensorflow/addons#693]
https://github.com/tensorflow/addons/blob/776b7511c1531f2cac3c822d30dc7bd7bcac3465/tensorflow_addons/losses/focal_loss.py#L132

**Comment by seanpmorgan** [comment]
cc @AakashKumarNain code owner

**Comment by AakashKumarNain** [comment]
@danijar any sample code to reproduce?

**Comment by danijar** [comment]
I just read through the code and wanted to highlight a potential problem. Looking at K.clip, it doesn't seem to provide a gradient for values that have been clipped. In this is indeed the case, a fix would be

```python
clipped = K.clip(y_pred, K.epsilon(), 1. - K.epsilon())
y_pred = y_pred - tf.stop_gradient(y_pred) + tf.stop_gradient(clipped)
```

**Comment by AakashKumarNain** [comment]
Thank you @danijar for looking into this and providing a fix. I will look into it once again and we will make the necessary code changes. Thank you again

**Comment by WindQAQ** [comment]
@AakashKumarNain Sorry for the late response. I'm fine with this. Maybe provide a testcase to verify the patch is needed!

**Comment by AakashKumarNain** [comment]
Sorry for the delay. I was on vacation. I will check and if needed, fix this during this weekend

**Comment by AakashKumarNain** [comment]
`K.clip` is just clip by value. It doesn't seem to affect the gradients, hence no fix is needed IMO. Though looking for any additional inputs

**Comment by danijar** [comment]
What do you mean? Clip by value is not differentiable.

**Comment by AakashKumarNain** [comment]
The same thing is being used here in `binary_crossentropy` calculations.
https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/backend.py#L4558-L4588

**Comment by Squadrick** [comment]
@danijar I understand the confusion. 

[`tf.clip_by_value`](https://github.com/tensorflow/tensorflow/blob/f7f4049ded8ad2bfbaecde1ff2d42460cc9fa762/tensorflow/python/ops/clip_ops.py#L38) uses `tf.minimum` and `tf.maximum` currently (ignoring the TODO). 

Both `tf.minimum` and `tf.maximum` have their gradients defined [here](https://github.com/tensorflow/tensorflow/blob/606a542c81df033d493d25e32b9f175b947fcbbb/tensorflow/core/ops/math_grad.cc#L617). So yes, as @AakashKumarNain mentioned, using `tf.clip_by_value` shouldn't be a problem. 

As for the gradient, it'll be 0 outside the range of `min_value` and `max_value`.

**Comment by danijar** [comment]
Exactly. Is it supposed to be 0 outside of the range?

**Comment by Squadrick** [comment]
@danijar My bad, you're right. A clip is used in BCE before taking a `log` for numerical stability. That won't be needed for focal loss.

Funnily enough in the `tf.keras` implementation of  BCE, that does both a clip followed by an `output + eps`. I think the addition of `epsilon` after the clip won't be needed.

**Comment by danijar** [comment]
My main point was that the gradient should probably be patched back onto the clipped tensor. Anyway, I'm not a user of the module.

**Comment by AakashKumarNain** [comment]
We will patch it.
